### DISCUSSION TITLE
Thinking about managing apps with modules

### DISCUSSION THREAD

**Discussion by aalaesar** [nextcloud/ansible-collection-nextcloud-admin#237] [Ideas]
Hello there!
I've been thinking about modules related to nc apps management, as i think they should be one of the easiest to start development.
I've been exploring the occ cli to see the extent of  the possibilities and make some module specs.
Here are my notes:, feel free to comment or give ideas.

# occ app modules

### list of apps, status (enabled, disabled)

`occ app:list --output=json`
opt: `--shipped=true` to list state only of shipped apps (interesting as a metadata)

#### system level
`occ config:list <appname>`

> note:  output is in json by default

#### user level

`occ user:setting <userid> <appname>`
**this should probably not be in the app generic module**

### module specs

A simple `name` optional parameter to return info for a specific application

The data returned should be things like

- **state** enabled/disabled/'not installed'
- **some boolean flags** about updates available or if the app is a shipped one
-  _maybe config keys on the system level_
- _maybe the path to the app folder_ - some may need it...

### installation and removal

pretty straight forward :
`occ app:install <appname>`

opts:
- keep-disabled
- force
- allow-unstable

`occ app:remove <appname>`
 opts:
 - keep-data

note : installing is possible with tgz archive, but this is an entire different process
Will do later

#### one app

no json output available
`occ app:update --showonly <appname>`

- result empty when no update and rc 0
- result 1 line = `<appname> new version available: <version>`

#### all apps

`occ app:update --showonly --all `

- result empty when no update and rc 0
- result 1 line per app = `<appname> new version available: <version>`

### update apps

For diff mode, if it can be implemented, gathering current  version before is required.

#### one app

`occ app:update <appname>`

return `<appname> new version available: <version>` and proceed to update
else return nothing and rc 0
return `<appname> updated` and rc 0
use option `--quiet` to suppress output

#### all apps

`occ app:update --all`
but because of the shell return, or the lack of (with `--quiet` opt),
the module should list updates before running any command

## configuring apps

should be used by the config module

**Comment by staticdev** [comment]
Looks very good. I was thinking of building an upgrade role on top of it.

**Comment by wiktor2200** [comment]
100% agree, that's very good idea to use `occ` to mange apps. I had the same idea in December, but haven't enough time to properly implement that :)
Cross posting my last comment here for not getting lost in closed issues: https://github.com/nextcloud/ansible-collection-nextcloud-admin/issues/171#issuecomment-1356507067

**Comment by aalaesar** [comment]
Ho. I've forgotten  to write  the app module specs but this should be something simple like

params:
  Name: the app id in the NC app store with maybe a special case for 'all'
  State: absent/present/latest
  Enable: true/false

**Comment by aalaesar** [comment]
hello there ! 
I was quite busy those few weeks but I finally got enough time to work on a draft for app_infos: 
#259 
it's pretty slow for fetching all apps I think, not sure why.
do you have any suggestion for it ?

Regards.
Aal

**Reply by wiktor2200** [reply]
Hi! From my experience `occ` commands are quite slow. What kind of slowness have you noticed? More or less seconds/minutes?

**Reply by aalaesar** [reply]
Hello there. Yes thats what I'm suspecting. I gave a detailed answer to @staticdev in the PR. Regarding that
